OpenshiftIT.tracingTest:70 HTTP status 403

[ERROR] Tests run: 1, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 16.559 s <<< FAILURE! - in me.snowdrop.istio.dt.OpenshiftIT
[ERROR] tracingTest  Time elapsed: 16.189 s  <<< FAILURE!
org.opentest4j.AssertionFailedError: 

expected: 200
 but was: 403
        at me.snowdrop.istio.dt.OpenshiftIT.tracingTest(OpenshiftIT.java:70)

[INFO] 
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   OpenshiftIT.tracingTest:70 
expected: 200
 but was: 403

https://github.com/snowdrop/istio-distributed-tracing-example/blob/a0b5ddbde27ffa02c698d1a486b45eb7d94df265/tests/src/test/java/me/snowdrop/istio/dt/OpenshiftIT.java#L70

--- a/README.adoc
+++ b/README.adoc
@@ -150,7 +150,6 @@ After everything is set up run the following commands.
 .Commands to execute the integration tests
 [source,bash]
 ----
-oc new-project <project-name>
 mvn clean verify -pl spring-boot-istio-distributed-tracing-cute-name-service -Popenshift -Ddekorate.deploy=true
 mvn clean verify -pl spring-boot-istio-distributed-tracing-greeting-service -Popenshift -Ddekorate.deploy=true
 oc create -f rules/greeting-gateway.yml

--- a/README.adoc
+++ b/README.adoc
@@ -71,11 +71,39 @@ oc delete project <your project name>
 
 To run integration tests, first follow the steps on the <<environment-preparation>> section to create and prepare a new namespace.
 
+Moreover, for the integration tests to pass, you need to disable the security of Jaeger UI first. By default, when installing the Red Hat OpenShift distributed tracing platform operator, it will install a Jaeger instance with the security enabled:
+
+[source,yaml]
+----
+  ingress:
+    enabled: true
+    openshift:
+      htpasswdFile: /etc/proxy/htpasswd/auth
+      sar: '{"namespace": "istio-system", "resource": "pods", "verb": "get"}'
+    options: {}
+    resources: {}
+    security: oauth-proxy
+----
+
+To disable, you need to update the Jaeger instance with:
+
+[source,yaml]
+----
+  ingress:
+    enabled: true
+    options: {}
+    resources: {}
+    security: none
+----
+
+And update it. The operator will automatically reload the Jaeger instance with the new configuration.
+
 After everything is set up run the following commands.
 
 .Commands to execute the integration tests
 [source,bash]
 ----
+oc new-project <project-name>
 mvn clean verify -pl spring-boot-istio-distributed-tracing-cute-name-service -Popenshift -Ddekorate.deploy=true
 mvn clean verify -pl spring-boot-istio-distributed-tracing-greeting-service -Popenshift -Ddekorate.deploy=true
 oc create -f rules/greeting-gateway.yml

--- a/README.adoc
+++ b/README.adoc
@@ -1,12 +1,15 @@
 = Istio Distributed Tracing Mission
+:icons: font
+:toc: left
+:toclevels: 2
 
 == Purpose
 Showcase Istio's Distributed Tracing via a (minimally) instrumented set of Spring Boot applications
 
 == Prerequisites
 . JDK 11+ installed with JAVA_HOME configured appropriately
-. Openshift 4.7 cluster
-. Istio installed on the aforementioned cluster using the link:https://docs.openshift.com/container-platform/4.11/service_mesh/v2x/installing-ossm.html[Red Hat OpenShift Service Mesh Operator].
+. Openshift 4.12 cluster
+. Istio installed on the aforementioned cluster using the link:https://docs.openshift.com/container-platform/4.12/service_mesh/v2x/installing-ossm.html[Red Hat OpenShift Service Mesh Operator].
 . Login to the cluster with the *admin* user
 
 [#environment-preparation]
@@ -18,7 +21,7 @@ Create a new project/namespace on the cluster. This is where your application wi
 oc new-project <whatever valid project name you want>
 ```
 
-Add the `<whatever valid project name you want>` project to the Istio member roll. More information in this link:https://docs.openshift.com/container-platform/4.11/service_mesh/v2x/ossm-create-mesh.html#ossm-about-adding-namespace_ossm-create-mesh[link].
+Add the `<whatever valid project name you want>` project to the Istio member roll. More information in this link:https://docs.openshift.com/container-platform/4.12/service_mesh/v2x/ossm-create-mesh.html#ossm-about-adding-namespace_ossm-create-mesh[link].
 
 == Build and deploy the application
 
@@ -71,8 +74,21 @@ oc delete project <your project name>
 
 To run integration tests, first follow the steps on the <<environment-preparation>> section to create and prepare a new namespace.
 
-Moreover, for the integration tests to pass, you need to disable the security of Jaeger UI first. By default, when installing the Red Hat OpenShift distributed tracing platform operator, it will install a Jaeger instance with the security enabled:
+=== Remove Jaeger UI security
 
+[CAUTION]
+====
+For the integration tests to work it is required that the Jaeger UI isn't secured by authentication, as it is in the default configuration.
+====
+
+[WARNING]
+=====
+Disabling security isn't the suggested configuration but it is required for the integration tests to work.
+=====
+
+By default, when installing the Red Hat OpenShift distributed tracing platform operator, it will install a Jaeger instance with the security enabled:
+
+.Secured Jaeger UI configuration
 [source,yaml]
 ----
   ingress:
@@ -85,8 +101,39 @@ Moreover, for the integration tests to pass, you need to disable the security of
     security: oauth-proxy
 ----
 
-To disable, you need to update the Jaeger instance with:
+Disabling security is accomplished by setting the `spec.ingress.security` on the Jaeger operator to `none`.
 
+This can be done either by using the OpenShift console (`Installed operators > Red Hat OpenShift distributed tracing platform > Jaguer > Edit Jaeger`) or using the CLI.
+
+[NOTE]
+====
+More information on the Jaeger configuration at the link:https://docs.openshift.com/container-platform/4.12/service_mesh/v2x/ossm-reference-jaeger.html[Jaeger configuration reference page].
+====
+
+To disable security execute the following CLI commands.
+
+.Set the `istio-system` as the default project
+[source,bash]
+----
+oc project istio-system
+----
+
+.Patch jaeger to disable console security
+[source,bash]
+----
+oc patch jaeger jaeger --patch '{"spec":{"ingress":{"security": "none" }}}' --type=merge
+----
+
+The result should be:
+
+[source]
+----
+jaeger.jaegertracing.io/jaeger patched
+----
+
+Which will result in the following configuration.
+
+.Unsecured Jaeger UI configuration
 [source,yaml]
 ----
   ingress:
@@ -96,7 +143,7 @@ To disable, you need to update the Jaeger instance with:
     security: none
 ----
 
-And update it. The operator will automatically reload the Jaeger instance with the new configuration.
+=== Execute the Integration Tests
 
 After everything is set up run the following commands.
 

--- a/pom.xml
+++ b/pom.xml
@@ -24,12 +24,13 @@
 
     <spring-cloud-sleuth.version>3.1.8</spring-cloud-sleuth.version>
     <spring-cloud-sleuth-otel.version>1.1.3</spring-cloud-sleuth-otel.version>
+    <opentelemetry-exporter-otlp.version>1.26.0</opentelemetry-exporter-otlp.version>
 
     <!-- Default values  / used when no profile is activated-->
     <greeting-service.port>8090</greeting-service.port>
     <cute-name-service.port>8091</cute-name-service.port>
     <cute-name-service.endpoint>localhost:8091</cute-name-service.endpoint>
-    <jaeger.endpoint>http://localhost:14268/api/traces</jaeger.endpoint>
+    <jaeger.endpoint>http://localhost:4317/api/traces</jaeger.endpoint>
   </properties>
   <licenses>
     <license>
@@ -105,6 +106,11 @@
         <groupId>org.springframework.cloud</groupId>
         <artifactId>spring-cloud-sleuth-otel-autoconfigure</artifactId>
         <version>${spring-cloud-sleuth-otel.version}</version>
+      </dependency>
+      <dependency>
+        <groupId>io.opentelemetry</groupId>
+        <artifactId>opentelemetry-exporter-otlp</artifactId>
+        <version>${opentelemetry-exporter-otlp.version}</version>
       </dependency>
       <dependency>
         <groupId>com.mks.api</groupId>

--- a/rules/greeting-gateway.yml
+++ b/rules/greeting-gateway.yml
@@ -25,7 +25,7 @@ spec:
   http:
   - match:
     - uri:
-        prefix: /greeting
+        prefix: /greeting/
     rewrite:
       uri: /
     route:

--- a/spring-boot-istio-distributed-tracing-cute-name-service/pom.xml
+++ b/spring-boot-istio-distributed-tracing-cute-name-service/pom.xml
@@ -25,6 +25,19 @@
       <groupId>org.springframework.boot</groupId>
       <artifactId>spring-boot-starter-actuator</artifactId>
     </dependency>
+    <dependency>
+      <groupId>org.springframework.cloud</groupId>
+      <artifactId>spring-cloud-starter-sleuth</artifactId>
+    </dependency>
+    <!-- Add OpenTelemetry tracer -->
+    <dependency>
+      <groupId>org.springframework.cloud</groupId>
+      <artifactId>spring-cloud-sleuth-otel-autoconfigure</artifactId>
+    </dependency>
+    <dependency>
+      <groupId>io.opentelemetry</groupId>
+      <artifactId>opentelemetry-exporter-otlp</artifactId>
+    </dependency>
   </dependencies>
 
   <build>
@@ -41,12 +54,8 @@
       <id>openshift</id>
       <properties>
         <!-- Override properties / used when deploying to Openshift -->
-        <greeting-service.port>8080</greeting-service.port>
         <cute-name-service.port>8080</cute-name-service.port>
-        <cute-name-service.endpoint>
-          spring-boot-istio-distributed-tracing-cute-name-service:8080
-        </cute-name-service.endpoint>
-        <jaeger.endpoint>http://jaeger-collector.istio-system.svc:14268/api/traces</jaeger.endpoint>
+        <jaeger.endpoint>http://jaeger-collector.istio-system:4317/api/traces</jaeger.endpoint>
       </properties>
       <dependencies>
         <dependency>

--- a/spring-boot-istio-distributed-tracing-cute-name-service/src/main/resources/application.yaml
+++ b/spring-boot-istio-distributed-tracing-cute-name-service/src/main/resources/application.yaml
@@ -1,8 +1,19 @@
 server.port: ${cute-name-service.port}
+spring:
+  application:
+    name: ${project.artifactId}
+  sleuth:
+    otel:
+      config:
+        # The trace-id-ratio-based property defines the sampling ratio for the spans collected.
+        # The value 1.0 means that all spans will be exported.
+        trace-id-ratio-based: 1.0
+      exporter:
+        otlp:
+          endpoint: ${jaeger.endpoint}
 # Dekorate
 dekorate:
   openshift:
     name: cute-name-service
-    expose: true
   s2i:
     builder-image: registry.access.redhat.com/ubi8/openjdk-11:1.14

--- a/spring-boot-istio-distributed-tracing-greeting-service/pom.xml
+++ b/spring-boot-istio-distributed-tracing-greeting-service/pom.xml
@@ -33,6 +33,10 @@
       <groupId>org.springframework.cloud</groupId>
       <artifactId>spring-cloud-sleuth-otel-autoconfigure</artifactId>
     </dependency>
+    <dependency>
+      <groupId>io.opentelemetry</groupId>
+      <artifactId>opentelemetry-exporter-otlp</artifactId>
+    </dependency>
   </dependencies>
 
   <build>
@@ -50,9 +54,8 @@
       <properties>
         <!-- Override properties / used when deploying to Openshift -->
         <greeting-service.port>8080</greeting-service.port>
-        <cute-name-service.port>8080</cute-name-service.port>
-        <cute-name-service.endpoint>cute-name-service:8080</cute-name-service.endpoint>
-        <jaeger.endpoint>http://jaeger-collector.istio-system.svc:14268/api/traces</jaeger.endpoint>
+        <cute-name-service.endpoint>cute-name-service:80</cute-name-service.endpoint>
+        <jaeger.endpoint>http://jaeger-collector.istio-system:4317/api/traces</jaeger.endpoint>
       </properties>
       <dependencies>
         <dependency>

--- a/spring-boot-istio-distributed-tracing-greeting-service/src/main/resources/application.yaml
+++ b/spring-boot-istio-distributed-tracing-greeting-service/src/main/resources/application.yaml
@@ -3,12 +3,15 @@ server.port: ${greeting-service.port}
 service:
   cute-name:
     name: ${cute-name-service.endpoint}
-
 spring:
   application:
     name: ${project.artifactId}
   sleuth:
     otel:
+      config:
+        # The trace-id-ratio-based property defines the sampling ratio for the spans collected.
+        # The value 1.0 means that all spans will be exported.
+        trace-id-ratio-based: 1.0
       exporter:
         otlp:
           endpoint: ${jaeger.endpoint}
@@ -17,6 +20,5 @@ spring:
 dekorate:
   openshift:
     name: greeting-service
-    expose: true
   s2i:
     builder-image: registry.access.redhat.com/ubi8/openjdk-11:1.14

--- a/tests/src/test/java/me/snowdrop/istio/dt/OpenshiftIT.java
+++ b/tests/src/test/java/me/snowdrop/istio/dt/OpenshiftIT.java
@@ -54,20 +54,18 @@ public class OpenshiftIT {
         RestAssured
                 .given()
                 .baseUri(ingressGatewayURL.toString())
-                .get("/greeting/api/greeting");
+                .get("/greeting/api/greeting")
+                .then().statusCode(200);
 
         TimeUnit.SECONDS.sleep(10); // wait 10 seconds so span will show up in the jaeger
 
-        final Response response =
-                RestAssured
-                .given()
+        final Response response = RestAssured.given()
                 .baseUri(jaegerQueryURL.toString())
                 .relaxedHTTPSValidation() //jaeger uses https, so we need to trust the cert
-                .param("service", ISTIO_INGRESS_GATEWAY_NAME)
+                .param("service", ISTIO_INGRESS_GATEWAY_NAME + "." + ISTIO_NAMESPACE)
                 .param("start", startTime)
-                .get("/api/traces");
-
-        assertThat(response.statusCode()).isEqualTo(200);
+                .get("/api/traces")
+                .then().statusCode(200).extract().response();
 
         final JsonPath jsonPath = response.jsonPath();
 

